clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -467,148 +483,178 @@ class MarkDeviceFunction : public RecursiveASTVisitor<MarkDeviceFunction> {
+      if (auto *FD = dyn_cast<FunctionDecl>(Record.Callee->getDecl()))
+        if (FD->hasBody() && FD->hasAttr<SYCLDeviceAttr>())
+          SemaRef.addSyclDeviceDecl(FD);
   }
 
-  // Traverses over CallGraph to collect list of attributes applied to
-  // functions called by SYCLKernel (either directly and indirectly) which needs
-  // to be propagated down to callers and applied to SYCL kernels.
-  // For example, reqd_work_group_size, vec_len_hint, reqd_sub_group_size
-  // Attributes applied to SYCLKernel are also included
-  // Returns the kernel body function found during traversal.
-  FunctionDecl *
-  CollectPossibleKernelAttributes(FunctionDecl *SYCLKernel,
-                                  llvm::SmallVector<Attr *, 4> &Attrs) {
-    typedef std::pair<FunctionDecl *, FunctionDecl *> ChildParentPair;
-    llvm::SmallPtrSet<FunctionDecl *, 16> Visited;
-    llvm::SmallVector<ChildParentPair, 16> WorkList;
-    WorkList.push_back({SYCLKernel, nullptr});
-    FunctionDecl *KernelBody = nullptr;
+  CallGraphNode *getNodeForKernel(FunctionDecl *Kernel) {
+    assert(CG.getNode(Kernel) && "No call graph entry for a kernel?");
+    return CG.getNode(Kernel);
+  }
 
-    while (!WorkList.empty()) {
-      FunctionDecl *FD = WorkList.back().first;
-      FunctionDecl *ParentFD = WorkList.back().second;
-
-      // To implement rounding-up of a parallel-for range the
-      // SYCL header implementation modifies the kernel call like this:
-      // auto Wrapper = [=](TransformedArgType Arg) {
-      //  if (Arg[0] >= NumWorkItems[0])
-      //    return;
-      //  Arg.set_allowed_range(NumWorkItems);
-      //  KernelFunc(Arg);
-      // };
-      //
-      // This transformation leads to a condition where a kernel body
-      // function becomes callable from a new kernel body function.
-      // Hence this test.
-      if ((ParentFD == KernelBody) && isSYCLKernelBodyFunction(FD))
-        KernelBody = FD;
+  void
+  AddSingleFunctions(const llvm::SmallPtrSetImpl<FunctionDecl *> &DevFuncs,
+                     const llvm::SmallPtrSetImpl<FunctionDecl *> &Recursive) {
+    DeviceFunctions.insert(DevFuncs.begin(), DevFuncs.end());
+    RecursiveFunctions.insert(Recursive.begin(), Recursive.end());
+  }
 
-      if ((ParentFD == SYCLKernel) && isSYCLKernelBodyFunction(FD)) {
-        assert(!KernelBody && "inconsistent call graph - only one kernel body "
-                              "function can be called");
-        KernelBody = FD;
-      }
+public:
+  DeviceFunctionTracker(Sema &S) : SemaRef(S) {
+    CG.addToCallGraph(S.getASTContext().getTranslationUnitDecl());
+    CollectSyclExternalFuncs();
+  }
 
-      WorkList.pop_back();
-      if (!Visited.insert(FD).second)
-        continue; // We've already seen this Decl
+  ~DeviceFunctionTracker() {
+    for (const FunctionDecl *FD : DeviceFunctions)
+      if (const FunctionDecl *Def = FD->getDefinition())
+        DiagDeviceFunction{SemaRef, RecursiveFunctions, Def->getBody()};
```

Comment:
Rather than create an object on every iteration of the loop (in a way that looks an awful lot like a function call, but with curly braces!), could we make a `DiagDeviceFunction` outside of the for loop and call a member function (passing in `Def->getBody()`) here instead?

Reply:
I thought that was the charm of this :)  Alright, I can do that.